writeToStream.js

```diff
@@ -173,8 +175,8 @@ function connect (opts, stream) {
   }
 
   // Password
-  if (password) {
-    if (password.length) {
+  if (password != null) {
+    if (providedUsername && (typeof password === 'string' || Buffer.isBuffer(password))) {
```

Comment:
I think it's possible to have a password without a username. So maybe this check should be relaxed. Can you check in the spec?
Even if it is not, it will be emitting "invalid password", which is not correct: it should be throwing something like "a username is required to use a password".

Reply:
From [v3.1.1](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html)

* [MQTT-3.1.2-20] If the Password Flag is set to 0, a password MUST NOT be present in the payload.
* [MQTT-3.1.2-21] If the Password Flag is set to 1, a password MUST be present in the payload.
* [MQTT-3.1.2-22] If the User Name Flag is set to 0, the Password Flag MUST be set to 0.

Reply:
I'll work on the error.

Reply:
Fantastic